ES6ify tailor-survey

SiAdcock opened the pull request:
## What does this change?

ES6-ifies the `tailor-survey` module. I prioritised this conversion as @oilnam would like to use `Set` in this module.

## What is the value of this and can you measure success?

✨ 

## Does this affect other platforms - Amp, Apps, etc?

Nah

## Tested in CODE?

Nah

<!-- AB test? https://git.io/v1V0x -->
<!-- AMP question? https://git.io/v1V0p -->
<!-- Does this PR meet the contributing guidelines? https://git.io/v1VEJ -->

PRBuilds commented:
PR build results:

screenshots
[mobile.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16573/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/mobile.png) • [wide.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16573/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/wide.png) • [tablet.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16573/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/tablet.png) • [desktop.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16573/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/desktop.png)
    
exceptions (0)
[thrown-exceptions.js](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16573/thrown-exceptions.js)    

webpagetest (3)
[performanceComparisonSummary.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-16573/performanceComparisonSummary.txt)

-automated message

regiskuckaertz commented on static/src/javascripts/projects/common/modules/experiments/tests/tailor-survey.js:
```diff
@@ -0,0 +1,263 @@
+    });
+};
+
+class TailorSurvey {
```
do we need a class? It looks like all the module has to produce is an immutable object. I'm in the process of writing a new test btw, and I will just do that, return `Object.freeze(spec)`.

Question: can I just create in my test in `javascript` (not `-legacy`) then?

SiAdcock replied:
Yeah, I don't necessarily agree with the current approach of creating AB test objects with the `new` keyword. You can totally create your AB test in `javascripts`, in fact I highly recommend that you do.

I don't want to change too much in this PR. Perhaps for future tests, we can recommend using the immutable object approach. Maybe I'll start by updating the docs.

regiskuckaertz replied:
fair enough 👍

gustavpursche commented on static/src/javascripts/projects/common/modules/experiments/tests/tailor-survey.js:
```diff
@@ -0,0 +1,263 @@
+    from 'raw-loader!common/views/experiments/tailor-survey.html';
+import ophan from 'ophan/ng';
+import template from 'lodash/utilities/template';
+import spaceFiller from 'common/modules/article/space-filler';
```
import { fillSpace } from '...' ?

gustavpursche commented on static/src/javascripts/projects/common/modules/experiments/tests/tailor-survey.js:
```diff
@@ -0,0 +1,263 @@
+
+const disableRadioButtons = buttonClassName => {
+    const radioButtons = document.getElementsByClassName(buttonClassName);
+    bonzo(radioButtons).each(button => {
```
`[...radioButtons].forEach()`?

GHaberis replied:
@gustavpursche I really like this way of iterating through html collections!

gustavpursche commented on static/src/javascripts/projects/common/modules/experiments/tests/tailor-survey.js:
```diff
@@ -0,0 +1,263 @@
+        'impressions-survey__content'
+    );
+
+    surveyContent[0].classList.add('js-impressions-survey__fadeout');
```
Why does flow doesn't say inything in this case? `surveyContent` is nullable ...

regiskuckaertz replied:
`getElementsByClassName : string => HTMLCollection`, so `length === 0` when nothing found 😃

regiskuckaertz replied:
so probably a good idea to `forEach` instead.

SiAdcock replied:
[Array access is unsafe](https://flow.org/en/docs/types/arrays/#toc-array-access-is-unsafe)

> In order to make this safe, Flow would have to mark every single array access as “possibly undefined”.
> Flow does not do this because it would be extremely inconvenient to use. You would be forced to refine the type of every value you get when accessing an array.

gustavpursche commented on static/src/javascripts/projects/common/modules/experiments/tests/tailor-survey.js:
```diff
@@ -0,0 +1,263 @@
+    ];
+
+    surveyQuestions.forEach(question => {
+        bean.on(question, 'click', event => {
```
While you are at it: could you convert this to use `addEventListener`?

sndrs replied:
given converting alone could allow errors to creep in, i think we need to be careful to limit conversion PRs to converting alone. it's a very good suggestion, but if it introduced a bug, it would be harder to pin down, given the breadth of changes already here (and that it does currently work). 

we can use `// #?` to flag things like this for review later though (even straight after merge, ideally), e.g. 
```js
       // #? use addEventListener 
       bean.on(question, 'click', event => {
```

to take it one step further, it might be a useful piece of work to remove bean as one job after this is done.

SiAdcock replied:
I agree with @sndrs (although I rushed in before I saw his comment and made this change anyway). There is a temptation to fix all the things, but we ought to show restraint, given the pace of refactoring.

regiskuckaertz replied:
Yes, definitely agreed 👍  I wouldn't want to have to disseminate those comments and see duplicates of the same idea coming up though -- what about having a single `low-hanging-fruits.md` file where to put them?

SiAdcock replied:
We've converted about 34 files and there are 10 occurrences of the `// #?` pattern with no duplicates yet. I like the idea of having the question right next to the code that needs to be looked at as it reduced the maintenance overhead. If we start to notice a lot of duplicate problems, then maybe we can rethink.

gustavpursche commented on static/src/javascripts/projects/common/modules/experiments/tests/tailor-survey.js:
```diff
@@ -0,0 +1,263 @@
+    canRun: () => boolean;
+
+    constructor() {
+        this.id = 'TailorSurvey';
```
What do you think about writing this as a pure Object

`const props = { id: '...' }` and assign it later as a whole to this `Object.assign(this, props)`? Less of the annoying and repetitive `this.` ..

SiAdcock replied:
As I mentioned to Regis higher up, I don't want to change too much in this PR. All our AB tests are exported as constructor functions. Introducing inconsistency as well as implementing ES6/Flow at this stage feels like a refactor too far.

SiAdcock commented:
I believe I've addressed the most pressing comments here (and added the points about AB tests to the backlog). Can I beg another quick review? @sndrs @gustavpursche @regiskuckaertz @GHaberis
